src/langchain_google_spanner/graph_store.py

```diff
@@ -0,0 +1,1176 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import datetime
+import re
+import string
+from abc import abstractmethod
+from typing import Any, Dict, List, Optional, Tuple, Type
+
+from google.cloud import spanner
+from google.cloud.spanner_v1 import param_types
+from langchain_community.graphs.graph_document import GraphDocument, Node, Relationship
+from langchain_community.graphs.graph_store import GraphStore
+from requests.structures import CaseInsensitiveDict
+
+
+def to_identifier(s: str) -> str:
+    return "`" + s + "`"
+
+
+def to_identifiers(s: List[str]) -> str:
+    return map(to_identifier, s)
+
+
+class NodeWrapper(object):
+    """Wrapper around Node to support set operations using node id"""
+
+    def __init__(self, node: Node):
+        self.node = node
+
+    def __hash__(self):
+        return hash(self.node.id)
+
+    def __eq__(self, other: Any):
+        if isinstance(other, NodeWrapper):
+            return self.node.id == other.node.id
+        return False
+
+
+class EdgeWrapper(object):
+    """Wrapper around Relationship to support set operations using edge source and target id"""
+
+    def __init__(self, edge: Relationship):
+        self.edge = edge
+
+    def __hash__(self):
+        return hash(
+            (
+                self.edge.source.id,
+                self.edge.target.id,
+            )
+        )
+
+    def __eq__(self, other: Any):
+        if isinstance(other, EdgeWrapper):
+            return (
+                self.edge.source.id == other.edge.source.id
+                and self.edge.target.id == other.edge.target.id
+            )
+        return False
+
+
+def partition_graph_docs(
+    graph_documents: List[GraphDocument],
+) -> Tuple(dict, dict):
+    """Returns nodes and edges grouped by the type.
+
+    Parameters:
+    - graph_documents: List[GraphDocument]
+
+    Returns:
+    - A tuple of two dictionaries. The first is nodes grouped by node types,
+      the second is edges grouped by edge types.
+    """
+    nodes = CaseInsensitiveDict()
+    edges = CaseInsensitiveDict()
+    for doc in graph_documents:
+        for node in doc.nodes:
+            s = nodes.setdefault(node.type, set())
+            w = NodeWrapper(node)
+            if w in s:
+                # Combine the properties for nodes with the same id.
+                n = next(filter(lambda v: v == w, s))
```

Comment:
This operation potentially be slow for big graphs, because it's iterating over a complete set which is not required as the `NodeWrapper` object is hashable & comparable based on their ids.

Reply:
Done!